File: Days 051 - 060/Day 59/FIleSyncAlgorithm.cs
using System;
using System.Collections.Generic;

namespace DailyCodingProblem
{
	internal class FileDirectory
	{
		private Dictionary<string, string> files = new Dictionary<string, string>();
		private readonly string name;

		public FileDirectory(string name)
		{
			this.name = name;
		}

		public bool AddFile(string path, string contents)
		{
			if (files.ContainsKey(path))
			{
				return false;
			}

			files.Add(path, contents);

			return true;
		}

		public bool UpdateFile(string path, string newContents)
		{
			if (!files.ContainsKey(path))
			{
				return false;
			}

			files[path] = newContents;

			return true;
		}

		public void SynchroniseTo(FileDirectory other)
		{
			List<string> filesToRemove = new List<string>();
			List<string> filesToUpdate = new List<string>();

			foreach (var file in files)
			{
				string path = file.Key;
				string contents = file.Value;

				if (!other.files.ContainsKey(path))
				{
					filesToRemove.Add(path);
				}
				else if (other.files[path] != files[path])
				{
					filesToUpdate.Add(path);
				}
			}

			foreach (string path in filesToRemove)
			{
				files.Remove(path);
			}

			foreach (string path in filesToUpdate)
			{
				files[path] = other.files[path];
			}

			foreach (var file in other.files)
			{
				string path = file.Key;
				string contents = file.Value;

				if (!files.ContainsKey(path))
				{
					files.Add(path, contents);
				}
			}
		}

		public void Print()
		{
			Console.WriteLine($"{name}\n");

			foreach (var file in files)
			{
				Console.WriteLine($"{file.Key}\t[{file.Value}]");
			}

			Console.WriteLine();
		}
	}

	internal class Day59
	{
		private static int Main(string[] args)
		{
			FileDirectory dirA = new FileDirectory("Directory A");
			dirA.AddFile("file1.txt", "file1 contents");
			dirA.AddFile("file2.txt", "file2's contents");
			dirA.AddFile("file3.txt", "file3 contents and some other stuff");

			FileDirectory dirB = new FileDirectory("Directory B");
			dirB.AddFile("file2.txt", "file2 contents");
			dirB.AddFile("file3.txt", "file3 contents");
			dirB.AddFile("file4.txt", "file4 contents");

			Console.WriteLine("BEFORE SYNCHRONISATION:");
			dirA.Print();
			dirB.Print();

			dirA.SynchroniseTo(dirB);

			Console.WriteLine("AFTER SYNCHRONISATION:");
			dirA.Print();
			dirB.Print();

			Console.ReadLine();

			return 0;
		}
	}
}

File: Days 051 - 060/Day 59/FileSyncAlgorithm.cpp
#include <iostream>
#include <map>
#include <string>
#include <vector>

class FileDirectory
{
private:
	std::map<std::string, std::string> files;
	std::string name;

public:
	explicit FileDirectory(const std::string& name)
		: name(name)
	{ }

	bool AddFile(const std::string& path, const std::string& contents)
	{
		if (files.find(path) != files.end())
		{
			return false;
		}

		files.insert(std::make_pair(path, contents));

		return true;
	}
	
	bool UpdateFile(const std::string& path, const std::string& newContents)
	{
		if (files.find(path) == files.end())
		{
			return false;
		}

		files[path] = newContents;

		return true;
	}

	void SynchroniseTo(FileDirectory other)
	{
		std::vector<std::string> filesToRemove;

		for (auto[path, contents] : files)
		{
			if (other.files.find(path) == other.files.end())
			{
				filesToRemove.push_back(path);
			}
			else if (other.files[path] != files[path])
			{
				files[path] = other.files[path];
			}
		}

		for (const std::string& path : filesToRemove)
		{
			files.erase(path);
		}

		for (auto[path, contents] : other.files)
		{
			if (files.find(path) == files.end())
			{
				files.insert(std::make_pair(path, contents));
			}
		}
	}

	void Print() const
	{
		std::cout << name << "\n\n";

		for (const auto[path, contents] : files)
		{
			std::cout << path << "\t[" << contents << "]\n";
		}

		std::cout << "\n";
	}
};

int main(int argc, char* argv[])
{
	FileDirectory dirA("Directory A");
	dirA.AddFile("file1.txt", "file1 contents");
	dirA.AddFile("file2.txt", "file2's contents");
	dirA.AddFile("file3.txt", "file3 contents and some other stuff");

	FileDirectory dirB("Directory B");
	dirB.AddFile("file2.txt", "file2 contents");
	dirB.AddFile("file3.txt", "file3 contents");
	dirB.AddFile("file4.txt", "file4 contents");

	std::cout << "BEFORE SYNCHRONISATION:\n";
	dirA.Print();
	dirB.Print();

	dirA.SynchroniseTo(dirB);

	std::cout << "AFTER SYNCHRONISATION:\n";
	dirA.Print();
	dirB.Print();

	std::cin.get();

	return 0;
}